### ISSUE TITLE
Strange behavior when using joblib parallel and numpyro effect handler for masking

### ISSUE THREAD

**Issue by vishu-tyagi** [pyro-ppl/numpyro#1855]
```python
import os

import numpy as np
import jax
from jax import random
import matplotlib.pyplot as plt
import seaborn as sns

import numpyro
import numpyro.distributions as dist
from numpyro.infer import NUTS, MCMC
import joblib
from joblib import Parallel, delayed

numpyro.enable_x64()
numpyro.set_platform("cpu")
numpyro.set_host_device_count(8)

# Set this to your build directory
BUILD_DIR = "/home/vishu/repro/"

assert numpyro.__version__ == "0.15.2"
assert jax.__version__ == "0.4.31"
assert joblib.__version__ == "1.4.2"


class Model():
    NAME = "model_without_mask"

    def __init__(self):
        self.rng_key = random.PRNGKey(0)
        self.build_dir = None
        self.mcmc_params = {
            "num_chains": 4,
            "num_warmup": 4000,
            "num_samples": 1000,
            "thinning": 1,
        }
        self.n_tests = None

    def _model(self, subject_ind, group_ind, y_obs=None):
        n_subjects = np.max(subject_ind) + 1
        n_groups = np.max(group_ind) + 1

        n_tests = self.n_tests
        n_data = subject_ind.shape[0]

        a_loc_loc_scale = numpyro.sample('a_loc_loc_scale', dist.HalfNormal(50))
        a_loc_scale = numpyro.sample('a_loc_scale', dist.HalfNormal(50))

        a_scale = numpyro.sample('a_scale', dist.HalfNormal(50))
        scale = numpyro.sample('scale', dist.HalfNormal(50))

        with numpyro.plate('n_tests', n_tests):
            a_loc_loc = numpyro.sample('a_loc_loc', dist.Normal(50., a_loc_loc_scale))

            with numpyro.plate("n_groups", n_groups):
                a_loc = numpyro.sample('a_loc', dist.Normal(a_loc_loc, a_loc_scale))

                with numpyro.plate("n_subjects", n_subjects):
                    a = numpyro.sample('a', dist.Normal(a_loc, a_scale))

        with numpyro.plate('n_tests', n_tests):
            with numpyro.plate('n_data', n_data):
                numpyro.sample(
                    'obs',
                    dist.Normal(a[subject_ind, group_ind], scale),
                    obs=y_obs
                )

    def run(
        self,
        subject_ind,
        group_ind,
        y_obs,
    ) -> tuple[MCMC, dict]:
        # Set up sampler
        sampler = NUTS(self._model)
        mcmc = MCMC(sampler, **self.mcmc_params)

        self.n_tests = y_obs.shape[-1]
        mcmc.run(self.rng_key, subject_ind, group_ind, y_obs=y_obs)
        posterior_samples = mcmc.get_samples()
        posterior_samples = {k: np.array(v) for k, v in posterior_samples.items()}
        return mcmc, posterior_samples


class MaskedModel(Model):
    NAME = "model_with_mask"

    def __init__(self):
        super(MaskedModel, self).__init__()

    def _model(self, subject_ind, group_ind, y_obs=None):
        n_subjects = np.max(subject_ind) + 1
        n_groups = np.max(group_ind) + 1

        n_tests = self.n_tests
        n_data = subject_ind.shape[0]

        mask = np.full((n_subjects, n_groups, n_tests), False)
        mask[subject_ind, group_ind, :] = True

        a_loc_loc_scale = numpyro.sample('a_loc_loc_scale', dist.HalfNormal(50))
        a_loc_scale = numpyro.sample('a_loc_scale', dist.HalfNormal(50))

        a_scale = numpyro.sample('a_scale', dist.HalfNormal(50))
        scale = numpyro.sample('scale', dist.HalfNormal(50))

        with numpyro.plate('n_tests', n_tests):
            a_loc_loc = numpyro.sample('a_loc_loc', dist.Normal(50., a_loc_loc_scale))

            with numpyro.plate("n_groups", n_groups):
                a_loc = numpyro.sample('a_loc', dist.Normal(a_loc_loc, a_loc_scale))

                with numpyro.plate("n_subjects", n_subjects):
                    with numpyro.handlers.mask(mask=mask):
                        a = numpyro.sample('a', dist.Normal(a_loc, a_scale))

        with numpyro.plate('n_tests', n_tests):
            with numpyro.plate('n_data', n_data):
                numpyro.sample(
                    'obs',
                    dist.Normal(a[subject_ind, group_ind], scale),
                    obs=y_obs
                )


def main(M, n_iters, n_jobs):

    # Body function for parallel execution using joblib
    def run(M, iter):
        # Read data
        subject_ind = np.array([
            0, 4, 1, 5, 2, 6, 3, 7, 0, 4, 1, 5, 2, 6, 3,
            7, 0, 4, 1, 5, 2, 6, 3, 7, 0, 4, 1, 5, 2, 6,
            3, 7, 0, 4, 1, 5, 2, 6, 3, 7, 0, 4, 1, 5, 2,
            6, 3, 7, 0, 4, 1, 5, 2, 6, 3, 7, 0, 4, 1, 5,
            2, 6, 3, 7, 0, 4, 1, 5, 2, 6, 3, 7, 0, 4, 1,
            5, 2, 6, 3, 7
        ])
        group_ind = np.array([
            0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1,
            0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1,
            0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1,
            0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1,
            0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1
        ])
        y_obs = np.array([
            [63.40434742953831, 43.14596633082837, 76.96001438498534, 71.27407171228093],
            [60.15631710392417, 45.3543617857226, 78.39989668718486, 65.61260585915308],
            [51.924780280679556, 53.46980346272739, 56.53150356377562, 66.11082373973957],
            [67.57259368289354, 40.83874964844216, 64.43140552664492, 51.01624553162892],
            [58.27719630945783, 54.99573497250754, 39.73317662830125, 85.26829691796175],
            [46.41546665448624, 60.26922685163903, 45.59614471676056, 67.50942281409614],
            [52.555758468803226, 53.842939757726704, 73.35559094895228, 64.94023703701232],
            [65.94910372719536, 43.284169131580086, 62.09931791280099, 76.5316358559518],
            [64.90908668680791, 47.45313434012833, 80.75190319175444, 67.43501981704397],
            [61.768931923290786, 46.29959115803473, 79.1692849705588, 66.72189405634309],
            [55.45284611333829, 54.75881637454442, 56.66908792972609, 67.15336208269588],
            [72.72436595843062, 40.67996221026947, 66.01744458829192, 50.333809458223804],
            [56.52409049868526, 54.62343747227834, 37.09319175426977, 83.51032962558446],
            [44.8897681854237, 60.459271177946206, 47.006409605538835, 65.01438054603192],
            [57.53656669079529, 56.70109473737409, 71.7919429643865, 64.72707216257841],
            [64.78316813301117, 44.93077343194879, 61.839832629951175, 84.02700515628605],
            [66.49880539110735, 47.71674195770745, 79.52723188108659, 70.55853451884187],
            [65.55911618572631, 46.05014576104088, 76.28481089891537, 65.4897710551513],
            [56.07435453712526, 55.052308187086936, 55.23940972840371, 65.39956170231807],
            [72.10210819105014, 43.378854360265514, 64.40879749321765, 49.31884576842487],
            [57.13403630604569, 57.294651552831716, 42.02237982383059, 81.13855900414062],
            [43.85069200532872, 62.5931096309739, 46.63511685335911, 64.5841718510329],
            [53.7867310657016, 58.956619008999695, 70.68728858250624, 60.92914353586672],
            [64.48409379064032, 41.85515178874887, 62.25547834093227, 77.4121725394366],
            [67.24065281126818, 47.5564366782129, 73.72491513909941, 71.47497886241683],
            [59.16113028746167, 48.891152382438825, 79.44516957301957, 68.38351380507883],
            [52.30538417898428, 53.558525010199894, 59.635418099791124, 64.97894307685499],
            [67.9148898493471, 43.36315008682732, 67.9546821383467, 51.313692810417976],
            [60.18552759841168, 57.551106225309, 40.092314427512605, 81.13334607318868],
            [46.14141909371866, 60.455127335771955, 48.654020100111666, 66.6419314953184],
            [51.96651084659327, 56.95462656289978, 69.11005824930547, 60.78685882025738],
            [62.7737010638273, 44.020871572138326, 62.35617862171528, 81.02652843325673],
            [65.24031894111546, 52.23378463477112, 80.99894346184136, 66.92648346130252],
            [60.9364631397759, 46.06140527697611, 78.01755274243389, 69.29405991664396],
            [54.65668583188604, 49.57272758283217, 54.97486492015479, 65.1006324179378],
            [70.92598994179527, 42.5063709084786, 67.51157095885462, 50.41040057895272],
            [56.93026243826525, 57.937891149378544, 34.77010081962167, 81.87468259689011],
            [42.93076409283401, 61.89100373655693, 47.369633653971654, 68.5087047816925],
            [52.752247042498546, 56.589659448655425, 68.69906584385959, 59.72720524234466],
            [67.91631469541781, 45.76201641272821, 59.72340124944221, 77.19312106563106],
            [68.69636831639536, 45.10523942671382, 80.5327453507377, 71.63491008924024],
            [61.37076088536876, 46.743553000009264, 78.07737865732307, 66.05219469624028],
            [55.41454480566086, 52.33876641429192, 59.01012705578315, 65.92662901811045],
            [68.57191502873061, 42.627911312577424, 66.46767302487983, 51.809247832341185],
            [56.70928909731914, 53.890238088656936, 37.204110244790186, 82.32985721614956],
            [51.982633202876606, 62.71712834991861, 48.9995343212048, 65.47489729391],
            [50.29875569480898, 55.66586813622932, 68.50343856591563, 65.2071370088477],
            [66.1091548706764, 44.152722103430385, 60.13536086184608, 83.64074825639389],
            [64.88098750716016, 46.536845116676425, 77.91728937327548, 70.87546315723642],
            [63.32467168700141, 45.68829946262205, 77.89252470424205, 66.78272333119334],
            [53.77399815158057, 53.35949971119939, 52.5560644658771, 65.14775735156857],
            [69.83522549115283, 45.30585953709808, 63.94546387001867, 53.01204840567921],
            [56.907738176379084, 54.26294774368162, 39.34333361226872, 81.99208094293077],
            [48.60112251379282, 59.9952488918378, 46.44540104635856, 66.7728596296786],
            [54.04018968024798, 55.873684943161535, 67.65605944692325, 61.64884973357708],
            [63.568527644600124, 43.491234875132974, 56.21869301773514, 76.07985137929543],
            [65.59974801423215, 49.07636881813318, 77.4145253805967, 69.85945618196496],
            [61.92715599653711, 44.85311889195673, 80.04741792036394, 67.1254380388301],
            [54.81086362670224, 53.93783242634863, 55.59892016190793, 67.74236166916727],
            [68.4019294744451, 40.53823494806909, 65.87708338954619, 55.1750647610908],
            [57.009585011705745, 57.46961972232486, 36.036660086101634, 82.572750060243],
            [42.70172101338347, 62.29743999803725, 43.377386054142576, 68.6530357080843],
            [55.88244086553588, 58.43046254666598, 74.06157753583783, 64.0410801405954],
            [64.63748670686944, 43.77882117824105, 63.006018933122505, 75.53164536996354],
            [65.50926783429668, 47.505398949231044, 79.86983988109931, 70.14297621325457],
            [63.12648361951137, 51.466266389609494, 72.78306084012087, 66.7640696817737],
            [58.021450034456684, 53.98271871370592, 54.0748900159514, 68.35912596551685],
            [68.55540306662691, 43.74493131466755, 63.578042822278185, 53.35022912677567],
            [57.339946651797646, 59.53514145647941, 40.463743105983994, 82.83366809620989],
            [47.924611933008066, 60.7122822868287, 42.49892658500698, 67.70665135418078],
            [52.46282086426171, 54.12689937520931, 72.16438985241172, 59.440168251355054],
            [68.23839616158186, 47.702633352976974, 59.965579978122896, 76.72688476829364],
            [63.45596925992426, 48.654467184276, 78.4789436121994, 71.82083953009804],
            [62.583596296833655, 44.55648941692456, 80.26196604594095, 66.17416537120064],
            [53.532392870768575, 59.20199800403784, 56.19005821881272, 68.06333048777485],
            [67.45590001170218, 44.35066089771819, 66.78716951157789, 52.61336233140038],
            [55.27360169096901, 57.80118855581109, 40.36366770673722, 83.56350772394782],
            [45.16236971570012, 64.94230473826953, 51.074932652915145, 70.50083218815999],
            [56.621272215581925, 54.740046370223176, 66.7287995482349, 62.306500859587004],
            [62.778162801511, 45.33136844186428, 59.769731867206644, 80.70174981555236]
        ])

        # # Check
        # _subject_ind, _group_ind, _y_obs = (
        #     np.load(os.path.join(BUILD_DIR, "subject_ind.npy")),
        #     np.load(os.path.join(BUILD_DIR, "group_ind.npy")),
        #     np.load(os.path.join(BUILD_DIR, "y_obs.npy")),
        # )
        # assert np.all(subject_ind == _subject_ind)
        # assert np.all(group_ind == _group_ind)
        # assert np.all(y_obs == _y_obs)

        # Initialize model
        model = M()
        model.build_dir = os.path.join(
            BUILD_DIR, f"iter_{iter}", model.NAME
        )
        os.makedirs(model.build_dir, exist_ok=True)

        # Run model
        mcmc, posterior_samples = model.run(
            subject_ind=subject_ind, group_ind=group_ind, y_obs=y_obs
        )
        print(mcmc.print_summary())

        # Save traceplot of a_loc
        a_loc = posterior_samples["a_loc"]

        nrows, ncols = a_loc.shape[1:]
        fig, axes = plt.subplots(nrows, ncols)
        for i in range(nrows):
            for j in range(ncols):
                ax = axes[i, j]
                for chain in range(4):
                    sns.lineplot(
                        x=np.arange(1000),
                        y=a_loc[chain * 1000: (chain + 1) * 1000, i, j],
                        ax=ax,
                        label=f"Chain {chain}",
                    )
                if ax.get_legend():
                    ax.get_legend().remove()

        axes[0, 0].legend()
        dest = os.path.join(model.build_dir, "a_loc.png")
        fig.savefig(dest)
        plt.close(fig)
        print(f"Saved to {dest}")
        return


    with Parallel(n_jobs=n_jobs) as parallel:
        parallel(
            delayed(run)(M, iter)
            for iter in range(n_iters)
        )


if __name__ == "__main__":
    os.makedirs(BUILD_DIR, exist_ok=True)

    # This results in poor rhats and trace plots
    # with n_jobs = -1
    M = MaskedModel
    n_jobs = -1

    # # This results in good rhats and trace plots
    # # after setting n_jobs = 1
    # M = MaskedModel
    # n_jobs = 1

    # # This results in good rhats and trace plots
    # # irrespective of value of n_jobs
    # M = Model
    # n_jobs = 1

    main(M=M, n_iters=1, n_jobs=n_jobs)


```

Repro:
Python version 3.11.9

requirements.txt
```
numpy==1.24.2
jax[cpu]==0.4.31
numpyro[cpu]==0.15.2
matplotlib==3.7.0
seaborn==0.12.2
joblib==1.4.2
```

Build the environment:
```
python3.11 -m venv .venv && \
source .venv/bin/activate && \
pip install --upgrade pip && \
pip install -r requirements.txt
```

Problem:
I was testing a simple hierarchical Bayesian model that models data belonging to two different groups of subjects, and the data for these two groups comes from multiple sites (indicated by n_tests attribute of the models, which equals 4 since `y_obs` has shape `(80, 4)`). The data that I'm using here was simulated (which is involved), so I'm just using one of the simulated datasets where this model broke.

I was interested in understanding the behavior of masking. Since a given subject belongs to exactly one group, there are a lot of redundant parameters in the site `a`, which is of shape `(8, 2, 4)` --  8 is the total number of subjects, 2 groups, 4 tests (corresponding to 4 sites). There are 4 subjects for each of the two groups.

If I run the `MaskedModel` with `n_jobs = -1`, it results in bad Rhats and a trace plot that looks like this for the `a_loc` parameter:

<img width="626" alt="Screenshot 2024-08-22 at 10 11 10 PM" src="https://github.com/user-attachments/assets/79daafcd-ab9a-425f-9489-6bc541603980">

and, if I run the same model with `n_jobs = 1`, it results in good Rhats
<img width="631" alt="Screenshot 2024-08-22 at 10 12 10 PM" src="https://github.com/user-attachments/assets/90326d8f-c537-4390-bbc3-d4fcfc78f20a">

Also, I found this behavior to be consistent with multiple simulated datasets.

Additionally, if I run the model `Model` (that has no masking), it works just fine irrespective of `n_jobs` value.

I'm posting here in case anyone is aware of this, but I'll try to debug later. Any help is appreciated.

**Comment by vishu-tyagi** [comment]
It looks like if we move `numpyro.enable_x64()` inside the body function `run`, this fixes it. Not sure if this is the expected way to use numpyro with joblib?

**Comment by fehiepsi** [comment]
`numpyro.enable_x64()` just calls `jax.config.update("jax_enable_x64", True)`. I guess this is the expected way to use jax with joblib.

**Comment by fehiepsi** [comment]
`numpyro.enable_x64()` just calls `jax.config.update("jax_enable_x64", True)`. I guess this is the expected way to use jax with joblib.
